x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotDirectory.java

```diff
@@ -562,8 +563,19 @@ public static Directory create(
             );
         }
 
-        final String repositoryName = SNAPSHOT_REPOSITORY_SETTING.get(indexSettings.getSettings());
-        final Repository repository = repositories.repository(repositoryName);
+        final Repository repository;
+        final String repositoryName;
+        if (SNAPSHOT_REPOSITORY_UUID_SETTING.exists(indexSettings.getSettings())) {
+            final String repositoryUuid = SNAPSHOT_REPOSITORY_UUID_SETTING.get(indexSettings.getSettings());
+            repository = repositories.repositoryByUuid(repositoryUuid, SNAPSHOT_REPOSITORY_NAME_SETTING.get(indexSettings.getSettings()));
+            repositoryName = repository.getMetadata().name();
+        } else {
+            // repository containing pre-7.12 snapshots has no UUID so we assume it matches by name
+            repositoryName = SNAPSHOT_REPOSITORY_NAME_SETTING.get(indexSettings.getSettings());
```

Comment:
Should we assert that `index.version.created` is before 7.12?

Reply:
That's not the case unfortunately. The repository is in BWC mode because it contains at least one pre-7.12 snapshot, but that old snapshot is not necessarily anything to do with the snapshot we are searching.